File: backend/app/bloklar.py
"""Ders bloğu desenleri.

Bir müfredat satırının haftalık saatinin gün içinde nasıl parçalanacağını
kullanıcı belirler: 5 saatlik bir ders "2+2+1" ya da "1+1+2+1" olabilir.
Desen boşsa saatler tek tek dağıtılır.
"""
from __future__ import annotations

AYRAC = "+"
EN_UZUN_BLOK = 8


class DesenHatasi(ValueError):
    """Desen okunamadı ya da haftalık saate uymuyor."""


def coz(desen: str | None, haftalik_saat: int) -> list[int]:
    """'2+2+1' → [2, 2, 1]. Boş desen tek saatlik bloklara açılır."""
    if not desen or not desen.strip():
        return [1] * haftalik_saat

    parcalar: list[int] = []
    for ham in desen.replace(",", AYRAC).replace(" ", AYRAC).split(AYRAC):
        ham = ham.strip()
        if not ham:
            continue
        if not ham.isdigit():
            raise DesenHatasi(
                f"'{ham}' bir sayı değil. Deseni 2+2+1 gibi yazın."
            )
        sayi = int(ham)
        if sayi < 1:
            raise DesenHatasi("Blok uzunluğu en az 1 olmalı.")
        if sayi > EN_UZUN_BLOK:
            raise DesenHatasi(f"Tek blok en fazla {EN_UZUN_BLOK} saat olabilir.")
        parcalar.append(sayi)

    if not parcalar:
        return [1] * haftalik_saat

    toplam = sum(parcalar)
    if toplam != haftalik_saat:
        raise DesenHatasi(
            f"Blokların toplamı {toplam} saat, haftalık ders saati ise "
            f"{haftalik_saat}. İkisi eşit olmalı."
        )
    return parcalar


def yaz(bloklar: list[int]) -> str:
    """[2, 2, 1] → '2+2+1'"""
    return AYRAC.join(str(b) for b in bloklar)


def duzenle(desen: str | None, haftalik_saat: int) -> str:
    """Deseni doğrular ve tek biçime getirir. Hatalıysa DesenHatasi yükseltir."""
    return yaz(coz(desen, haftalik_saat))


def onerilenler(haftalik_saat: int) -> list[str]:
    """Arayüzde hızlı seçim için birkaç makul desen."""
    if haftalik_saat < 1:
        return []
    if haftalik_saat == 1:
        return ["1"]

    oneriler: list[str] = [yaz([1] * haftalik_saat)]

    # Mümkün olduğunca çift blok, kalan tek saat sona.
    ciftli, kalan = divmod(haftalik_saat, 2)
    oneriler.append(yaz([2] * ciftli + ([1] if kalan else [])))

    # Bir tane çift, gerisi tek.
    if haftalik_saat >= 3:
        oneriler.append(yaz([2] + [1] * (haftalik_saat - 2)))

    # Üçlü blok, uzun derslerde.
    if haftalik_saat >= 4:
        ucluler, art = divmod(haftalik_saat, 3)
        kuyruk = [2] if art == 2 else [1] * art
        oneriler.append(yaz([3] * ucluler + kuyruk))

    benzersiz: list[str] = []
    for o in oneriler:
        if o not in benzersiz:
            benzersiz.append(o)
    return benzersiz

File: frontend/src/lib/bloklar.ts
/** Ders bloğu desenleri. Backend'deki `app/bloklar.py` ile aynı kuralları uygular.
 *
 *  5 saatlik bir ders "2+2+1" ya da "1+1+2+1" olarak dağıtılabilir.
 *  Desen boşsa saatler tek tek dağıtılır.
 */

export const EN_UZUN_BLOK = 8;

export interface DesenSonucu {
  bloklar: number[];
  toplam: number;
  gecerli: boolean;
  /** Geçersizse kullanıcıya gösterilecek sebep. */
  hata: string | null;
}

/** Deseni çözer ve haftalık saate uyup uymadığını söyler. */
export function desenCoz(desen: string, haftalikSaat: number): DesenSonucu {
  const ham = desen.trim();
  if (!ham) {
    return {
      bloklar: Array(Math.max(0, haftalikSaat)).fill(1),
      toplam: haftalikSaat,
      gecerli: true,
      hata: null,
    };
  }

  const bloklar: number[] = [];
  for (const parca of ham.split(/[+,\s]+/).filter(Boolean)) {
    if (!/^\d+$/.test(parca)) {
      return { bloklar: [], toplam: 0, gecerli: false,
        hata: `"${parca}" bir sayı değil. Deseni 2+2+1 gibi yazın.` };
    }
    const sayi = Number(parca);
    if (sayi < 1) {
      return { bloklar: [], toplam: 0, gecerli: false, hata: "Blok uzunluğu en az 1 olmalı." };
    }
    if (sayi > EN_UZUN_BLOK) {
      return { bloklar: [], toplam: 0, gecerli: false,
        hata: `Tek blok en fazla ${EN_UZUN_BLOK} saat olabilir.` };
    }
    bloklar.push(sayi);
  }

  const toplam = bloklar.reduce((t, b) => t + b, 0);
  return {
    bloklar,
    toplam,
    gecerli: toplam === haftalikSaat,
    hata:
      toplam === haftalikSaat
        ? null
        : `Blokların toplamı ${toplam} saat, haftalık ders saati ise ${haftalikSaat}.`,
  };
}

const yaz = (bloklar: number[]) => bloklar.join("+");

/** Haftalık saate uyan birkaç makul desen — arayüzde hızlı seçim için. */
export function desenOnerileri(haftalikSaat: number): string[] {
  if (haftalikSaat < 1) return [];
  if (haftalikSaat === 1) return ["1"];

  const oneriler = [yaz(Array(haftalikSaat).fill(1))];

  const ciftli = Math.floor(haftalikSaat / 2);
  const kalan = haftalikSaat % 2;
  oneriler.push(yaz([...Array(ciftli).fill(2), ...(kalan ? [1] : [])]));

  if (haftalikSaat >= 3) {
    oneriler.push(yaz([2, ...Array(haftalikSaat - 2).fill(1)]));
  }
  if (haftalikSaat >= 4) {
    const ucluler = Math.floor(haftalikSaat / 3);
    const art = haftalikSaat % 3;
    const kuyruk = art === 2 ? [2] : Array(art).fill(1);
    oneriler.push(yaz([...Array(ucluler).fill(3), ...kuyruk]));
  }

  return [...new Set(oneriler)];
}

/** Tabloda gösterim: "1+1+1" gibi tek saatlik desenleri "tek" diye yazar. */
export function desenEtiketi(desen: string, haftalikSaat: number): string {
  const { bloklar } = desenCoz(desen, haftalikSaat);
  return bloklar.every((b) => b === 1) ? "tek saat" : bloklar.join("+");
}
